### PR TITLE
Fixes #25655 - Recurring Logic skips the first schedule before starting

### PR DESCRIPTION
**Pull request by sjha4**
In the method start_after we call next_occurence_time which calls @parser.next(start_time) when starting even if the start time is valid for the cron and we could very well start at the start-time itself. 
For example
cron = "* * * * *"
start_time = 12:35:00
We can start at 12:35 instead of starting at 12:36.

### CODE REVIEW

**Comment by beav** [comment]
@sjha4 without this PR, if I had a cron for `0 * * * *` and I entered it at 11:45, would the cron not fire at 12:00?

**Comment by adamruzicka** [comment]
@beav If I got that right if you entered a cron for `0 * * * *` and configured it to start at 12:00, it wouldn't fire until 13:00

**Comment by adamruzicka** [comment]
Depends whether you set the start at property, without it it should work fine

**Comment by beav** [comment]
test was successful. repro steps:

* create an hourly sync plan, set "start time" to something (I usually used ~5 min in the future)
* note that "next sync" one hour *past* the start time
* wait 6 minutes, note that sync plan didn't execute

with patch:

* create an hourly sync plan, set "start time" to something (again, ~5 min for my test)
* note that "next sync" is now correct
* wait 6 minutes, sync plan executes

File: app/models/foreman_tasks/recurring_logic.rb [hunk 1]
```diff
@@ -83,6 +83,7 @@ def cancel
 
     def next_occurrence_time(time = Time.zone.now)
       @parser ||= CronParser.new(cron_line, Time.zone)
+      return @parser.next(@parser.last(time)) if @parser.next(@parser.last(time)) >= time && tasks.count == 0
```
Comment (beav) [inline app/models/foreman_tasks/recurring_logic.rb]: this logic works and is correct, but is a little hard to follow. What data is `@parser.next(@parser.last(time))` capturing? is it the next run, but backdated one run?

Comment (sjha4) [reply app/models/foreman_tasks/recurring_logic.rb]: So with a start_at 12:46 if we have a cron "46 * * * *" , @parser.next(12:46) returns 13:46 while parser.last returns 11:46 and parser.next(11:46) returns 12:46 which is when we want the task to run..

Comment (beav) [reply app/models/foreman_tasks/recurring_logic.rb]: Can you add a comment saying something like "CronParser's `next` is not inclusive of the current time, so we step backwards one run to calculate the next run"? My concern is that if I came across this line in the future, I would not know what it was doing at all.

File: app/models/foreman_tasks/recurring_logic.rb [hunk 2]
```diff
@@ -83,6 +83,8 @@ def cancel
 
     def next_occurrence_time(time = Time.zone.now)
       @parser ||= CronParser.new(cron_line, Time.zone)
+      # @parser.next(start_time) is not inclusive of the start_time hence stepping back one run to include checking start_time for the first run.
```
Comment (beav) [inline app/models/foreman_tasks/recurring_logic.rb]: :+1:

**Comment by sjha4** [comment]
@adamruzicka : I am not sure if the tests ran ok..

**Comment by sjha4** [comment]
@adamruzicka : Does this look good to go to you?

File: app/models/foreman_tasks/recurring_logic.rb (hunk 1 above)
Comment (adamruzicka) [reply app/models/foreman_tasks/recurring_logic.rb]: One nitpick, it would be nice if you could avoid generating the time twice. Something like this might work
```ruby
before_next = @parser.next(@parser.last(time))
return before_next if before_next >= time && task.count.zero?
@parser.next(time)
```

**Comment by sjha4** [comment]
@adamruzicka : I updated the code as per your reviews..I am not sure if the test suite failure is related.

**Comment by sjha4** [comment]
Thanks @adamruzicka ! This needs to land with some version of katello 3.10..I am not sure what the dates would be but probably will need a foreman-tasks release when merged?

**Comment by beav** [comment]
@sjha4 do you mean like a 3.10.1?

**Comment by sjha4** [comment]
@beav: Yes..I thought 3.10.1 was out but turns out it isn't :smile: ..so that is the target release.

**Comment by adamruzicka** [comment]
Thanks @sjha4 !
